### PR TITLE
Add limit in analytical statement and analytical entries create

### PR DESCRIPTION
**Pull request by k9845**
Fixes:
- #737 Analysis Module: Limit entries in analytical statement to 50
- #736 Analysis Module: Limit analytical statements to 30 per Pillar Analysis

### CODE REVIEW

File: apps/analysis/tests/test_apis.py [hunk 1]
```diff
@@ -116,6 +122,82 @@ def test_create_analytical_statement(self):
...
+        project.add_member(user)
+        entry = self.create(Entry)
+        analysis = self.create(Analysis, project=project)
+        pillar = self.create(AnalysisPillar, analysis=analysis)
+        analytical_statement1 = self.create(AnalyticalStatement, analysis_pillar=pillar)
+        with self.assertRaises(ValidationError):
+            for _ in range(ANALYTICAL_STATEMENT_COUNT + 1):
+                self.create(AnalyticalStatement, analysis_pillar=pillar)
+        new_count = AnalyticalStatement.objects.filter(analysis_pillar=pillar).count()
+        self.assertEqual(new_count, ANALYTICAL_STATEMENT_COUNT)
```
Comment (annshress) [inline apps/analysis/tests/test_apis.py]: Lets remove this.

And add a test where client tries calling the api with more than `ANALYTICAL_STATEMENT_COUNT` statements for a given pillar.

File: apps/analysis/tests/test_apis.py [hunk 2]
```diff
@@ -116,6 +122,82 @@ def test_create_analytical_statement(self):
...
+        analytical_statement = self.create(AnalyticalStatement, analysis_pillar=pillar)
+        analytical_statement_entry = self.create(AnalyticalStatementEntry,
+                                                 analytical_statement=analytical_statement,
+                                                 entry=entry)
+        with self.assertRaises(ValidationError):
+            for _ in range(ANALYTICAL_ENTRIES_COUNT + 1):
+                self.create(AnalyticalStatementEntry,
+                            analytical_statement=analytical_statement)
+        new_count = AnalyticalStatementEntry.objects.filter(analytical_statement=analytical_statement).count()
+        self.assertEqual(new_count, ANALYTICAL_ENTRIES_COUNT)
```
Comment (annshress) [inline apps/analysis/tests/test_apis.py]: Same as for analytical statements.

File: apps/analysis/models.py [hunk 3]
```diff
@@ -2,7 +2,12 @@
 
 from django.db import models
 from django.contrib.postgres.fields import JSONField
+from django.core.exceptions import ValidationError
 
+from deep.settings import (
+    ANALYTICAL_ENTRIES_COUNT,
+    ANALYTICAL_STATEMENT_COUNT
+)
```
Comment (thenav56) [inline apps/analysis/models.py]: Unused imports

File: apps/analysis/serializers.py [hunk 4]
```diff
@@ -9,6 +9,10 @@
     NestedCreateMixin,
     NestedUpdateMixin
 )
+from deep.settings import (
+    ANALYTICAL_STATEMENT_COUNT,
+    ANALYTICAL_ENTRIES_COUNT
+)
```
Comment (thenav56) [inline apps/analysis/serializers.py]: Never use settings variables directly. Always use django.conf.settings.

```python
from django.conf import settings

settings.ANALYTICAL_STATEMENT_COUNT
....
```

File: apps/analysis/serializers.py [hunk 5]
```diff
@@ -17,65 +21,82 @@
...
-        analysis_id = self.context['view'].kwargs.get('analysis_id', None)
-        if analysis_id:
-            data['analysis_id'] = int(analysis_id)
+        analysis_pillar_id = self.context['view'].kwargs.get('analysis_pillar_id', None)
+        if analysis_pillar_id:
+            data['analysis_pillar_id'] = int(analysis_pillar_id)
+        # Validate the analytical_entries
+        analytical_entries_present = AnalyticalStatementEntry.objects.filter(analytical_statement=self.instance).count()
+        if analytical_entries_present + len(data.get('analyticalstatemententry_set', [])) >= ANALYTICAL_ENTRIES_COUNT:
+            raise serializers.ValidationError(f'Analytical entires count must be less than {ANALYTICAL_ENTRIES_COUNT}')
```
Comment (thenav56) [inline apps/analysis/serializers.py]: This should be enough. As entries all(PUT/POST) and none(PATCH) are send through the serializer.

```python3
entries = data.get('analyticalstatemententry_set')

if entries and len(entries) > settings.ANALYTICAL_ENTRIES_COUNT:
    raise serializers.ValidationError(f'Analytical entires count must be less than {ANALYTICAL_ENTRIES_COUNT}')
```

File: apps/analysis/serializers.py [hunk 6]
```diff
@@ -17,65 +21,82 @@
...
 
     def validate(self, data):
-        data['project_id'] = int(self.context['view'].kwargs['project_id'])
+        analysis_id = self.context['view'].kwargs.get('analysis_id', None)
+        if analysis_id:
+            data['analysis_id'] = int(analysis_id)
+        # validate analysis_statement
+        analytical_statement_present = AnalyticalStatement.objects.filter(analysis_pillar=self.instance).count()
+        if analytical_statement_present + len(data.get('analyticalstatement_set', [])) >= ANALYTICAL_STATEMENT_COUNT:
+            raise serializers.ValidationError(f'Analytical statement count must be less than {ANALYTICAL_STATEMENT_COUNT}')
```
Comment (thenav56) [inline apps/analysis/serializers.py]: Same logic here as well.

File: apps/analysis/tests/test_apis.py [hunk 7]
```diff
@@ -112,10 +205,72 @@ def test_create_analytical_statement(self):
...
+                {
+                    "statement": "coffee",
+                    "order": 1,
+                    "analytical_entries": [
+                        {
+                            "order": 1,
+                            "entry": entry.id,
+                        }
+                    ]
+                } for _ in range(ANALYTICAL_STATEMENT_COUNT)
```
Comment (thenav56) [inline apps/analysis/tests/test_apis.py]: Let's add test w/wo +1 for 400 and 201 responses.

```suggestion
                } for _ in range(ANALYTICAL_STATEMENT_COUNT + 1)
```

File: apps/analysis/tests/test_apis.py [hunk 8]
```diff
@@ -112,10 +205,72 @@ def test_create_analytical_statement(self):
...
+        self.authenticate(user)
+        response = self.client.post(url, data)
+        self.assert_400(response)
+        self.assertIn('non_field_errors', response.data['errors'])
+
+    def test_create_analytical_entries_greater_than_50_api_level(self):
+        user = self.create_user()
+        project = self.create_project()
+        project.add_member(user)
+        entries_list = [self.create(Entry) for _ in range(ANALYTICAL_ENTRIES_COUNT)]
```
Comment (thenav56) [inline apps/analysis/tests/test_apis.py]: Same here as well.

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/analysis/tests/test_apis.py (hunk 8 above)
Comment (k9845) [reply apps/analysis/tests/test_apis.py]: Done

File: apps/analysis/tests/test_apis.py (hunk 7 above)
Comment (k9845) [reply apps/analysis/tests/test_apis.py]: Done

File: apps/analysis/serializers.py (hunk 6 above)
Comment (k9845) [reply apps/analysis/serializers.py]: Done

File: apps/analysis/serializers.py (hunk 5 above)
Comment (k9845) [reply apps/analysis/serializers.py]: Done

File: apps/analysis/serializers.py (hunk 4 above)
Comment (k9845) [reply apps/analysis/serializers.py]: Done

File: apps/analysis/tests/test_apis.py (hunk 1 above)
Comment (k9845) [reply apps/analysis/tests/test_apis.py]: Done

File: apps/analysis/serializers.py [hunk 9]
```diff
@@ -17,65 +18,84 @@
...
-        if analysis_id:
-            data['analysis_id'] = int(analysis_id)
+        analysis_pillar_id = self.context['view'].kwargs.get('analysis_pillar_id', None)
+        if analysis_pillar_id:
+            data['analysis_pillar_id'] = int(analysis_pillar_id)
+        # Validate the analytical_entries
+        entries = data.get('analyticalstatemententry_set')
+        if entries and len(entries) >= settings.ANALYTICAL_ENTRIES_COUNT:
+            raise serializers.ValidationError(f'Analytical entires count must be less than\
+                                              {settings.ANALYTICAL_ENTRIES_COUNT}')
```
Comment (thenav56) [inline apps/analysis/serializers.py]: Let's do this.
- We don't want count to be > 50 not >= 50.
- If we don't have to break a string, let's not break, just move it to a new line.

```suggestion
        if entries and len(entries) > settings.ANALYTICAL_ENTRIES_COUNT:
            raise serializers.ValidationError(
                f'Analytical entires count must be less than {settings.ANALYTICAL_ENTRIES_COUNT}'
             )
```

File: apps/analysis/serializers.py (hunk 5 above)
Comment (thenav56) [reply apps/analysis/serializers.py]: Let's use `>` not `>=` as we want to have a count up to 50, not 49.

File: apps/analysis/serializers.py (hunk 6 above)
Comment (thenav56) [reply apps/analysis/serializers.py]: Same here as well.

File: apps/analysis/tests/test_apis.py (hunk 7 above)
Comment (thenav56) [reply apps/analysis/tests/test_apis.py]: Not done. Let's use w/wo + 1.

File: apps/analysis/tests/test_apis.py (hunk 8 above)
Comment (thenav56) [reply apps/analysis/tests/test_apis.py]: Same here as well.

File: apps/analysis/serializers.py [hunk 10]
```diff
@@ -9,6 +9,7 @@
     NestedCreateMixin,
     NestedUpdateMixin
 )
+from django.conf import settings
```
Comment (thenav56) [inline apps/analysis/serializers.py]: Let's move this before `from rest_framework...`

Basically also follow this while importing.

```
import native modules like datetime, json, csv

import installed modules like django, rest_framework

import local modules absolute path like from leads.models import Lead

import relative modules like from .models import Lead
```

Comment (k9845) [reply apps/analysis/serializers.py]: great!!!

**Review by thenav56 — APPROVED** [review]

**Review by annshress — APPROVED** [review]

**Review by annshress — APPROVED** [review]

**Pull request merged**